File: minegame-server/src/main/java/top/nipuru/minegame/server/game/friend/FriendManager.java
package top.nipuru.minegame.server.game.friend;

import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
import lombok.extern.slf4j.Slf4j;
import top.nipuru.minegame.common.message.database.QueryPlayerRequest;
import top.nipuru.minegame.server.TimeMgr;
import top.nipuru.minegame.server.game.player.BaseManager;
import top.nipuru.minegame.server.game.player.GamePlayer;
import top.nipuru.minegame.server.game.player.DataInfo;
import top.nipuru.minegame.server.util.GsonHolder;

import java.util.Collection;
import java.util.Collections;

@Slf4j
public class FriendManager extends BaseManager {

    private static final String moduleName = "friendship";

    /** 好友列表 */
    private final Int2ObjectMap<FriendshipData> friendships = new Int2ObjectOpenHashMap<>();
    /** 收到的好友请求列表 */
    private final Int2ObjectMap<FriendRequest> friendRequests = new Int2ObjectOpenHashMap<>();

    public FriendManager(GamePlayer player) {
        super(player);
    }

    public static void preload(QueryPlayerRequest request) throws Exception {
        preload(request, FriendshipData.class);
        preload(request, FriendRequest.class);
    }

    public void unpack(DataInfo dataInfo) throws Exception {
        for (FriendshipData friendship : dataInfo.unpackList(FriendshipData.class)) {
            friendships.put(friendship.friendId(), friendship);
        }
        for (FriendRequest friendRequest : dataInfo.unpackList(FriendRequest.class)) {
            friendRequests.put(friendRequest.friendId(), friendRequest);
        }
    }

    public void pack(DataInfo dataInfo) throws Exception {
        for (FriendshipData friendship : friendships.values()) {
            dataInfo.pack(friendship);
        }
        for (FriendRequest friendRequest : friendRequests.values()) {
            dataInfo.pack(friendRequest);
        }
    }

    public void init() {
        player.registerOfflineHandler(moduleName, this::handleOfflineData);
    }

    public Collection<FriendshipData> getFriends() {
        return Collections.unmodifiableCollection(friendships.values());
    }

    public boolean isFriend(int friendId) {
        return friendships.containsKey(friendId);
    }

    public FriendshipData getFriend(int friendId) {
        return friendships.get(friendId);
    }

    public int getFriendCount() {
        return friendships.size();
    }

    public Collection<FriendRequest> getFriendRequests() {
        return Collections.unmodifiableCollection(friendRequests.values());
    }

    public boolean hasFriendRequest(int friendId) {
        return friendRequests.containsKey(friendId);
    }

    public FriendRequest getFriendRequest(int playerId) {
        return friendRequests.get(playerId);
    }

    public void deleteFriend(int friendId, int friendDbId) {
        FriendshipData friendship = friendships.remove(friendId);
        if (friendship == null) return;
        player.delete(friendship);
        pushOfflineData(friendId, friendDbId, FriendshipOfflineData.DELETE, player.getPlayerId(), player.getDbId(), 0L);
    }

    public void rejectFriend(int friendId) {
        FriendRequest friendRequest = friendRequests.remove(friendId);
        if (friendRequest == null) return;
        player.delete(friendRequest);
    }

    public void acceptFriend(int friendId, int friendDbId) {
        FriendRequest friendRequest = friendRequests.remove(friendId);
        if (friendRequest == null) return;
        player.delete(friendRequest);
        FriendshipData friendship = friendships.get(friendId);
        if (friendship != null) return;
        friendship = new FriendshipData();
        friendship.friendId = friendId;
        friendship.createTime = TimeMgr.now();
        friendships.put(friendId, friendship);
        player.insert(friendship);
        pushOfflineData(friendId, friendDbId, FriendshipOfflineData.ACCEPT, player.getPlayerId(), player.getDbId(), friendship.createTime);
    }

    public void addFriend(int friendId, int friendDbId) {
        if (friendships.containsKey(friendId)) return;
        pushOfflineData(friendId, friendDbId, FriendshipOfflineData.REQUEST, player.getPlayerId(), player.getDbId(), TimeMgr.now());
    }

    // 通知请求不要在这里处理，要单独发
    private boolean handleOfflineData(String json, boolean isOnline) {
        FriendshipOfflineData data = GsonHolder.GSON.fromJson(json, FriendshipOfflineData.class);
        switch (data.cmd) {
            // 对方发送好友请求 新增一条未处理好友
            case FriendshipOfflineData.REQUEST -> {
                if (friendRequests.containsKey(data.friendId)) return true;
                FriendRequest friendRequest = new FriendRequest();
                friendRequest.friendId = data.friendId;
                friendRequest.createTime = data.createTime;
                friendRequests.put(friendRequest.friendId, friendRequest);
                player.insert(friendRequest);
            }
            // 对方接受好友请求 新增一条好友
            case FriendshipOfflineData.ACCEPT -> {
                FriendshipData friendship = friendships.get(data.friendId);
                if (friendship != null) return true;
                friendship = new FriendshipData();
                friendship.friendId = data.friendId;
                friendship.createTime = data.createTime;
                friendships.put(friendship.friendId, friendship);
                player.insert(friendship);
            }
            // 对方删除好友
            case FriendshipOfflineData.DELETE -> {
                FriendshipData friendship = friendships.remove(data.friendId);
                if (friendship == null) return true;
                player.delete(friendship);
            }
            default -> {
                return false;
            }
        }
        return true;
    }

    private void pushOfflineData(int friendId, int friendDbId, int cmd, int playerId, int dbId, long createTime) {
        FriendshipOfflineData offlineData = new FriendshipOfflineData(cmd, playerId, dbId, createTime);
        String json = GsonHolder.GSON.toJson(offlineData);
        player.pushOfflineDataTo(friendId, friendDbId, moduleName, json);
    }
}

File: minegame-game/src/main/kotlin/top/nipuru/minegame/game/gameplay/friend/FriendManager.kt
package top.nipuru.minegame.game.gameplay.friend

import it.unimi.dsi.fastutil.ints.Int2ObjectMap
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap
import top.nipuru.minegame.common.message.database.QueryPlayerRequest
import top.nipuru.minegame.game.gameplay.player.BaseManager
import top.nipuru.minegame.game.gameplay.player.DataInfo
import top.nipuru.minegame.game.gameplay.player.GamePlayer
import top.nipuru.minegame.game.gameplay.player.preload
import top.nipuru.minegame.game.gson
import top.nipuru.minegame.game.now
import java.util.*

private const val moduleName = "friendship"

class FriendManager(player: GamePlayer) : BaseManager(player) {
    /** 好友列表  */
    private val friendships: Int2ObjectMap<FriendshipData> = Int2ObjectOpenHashMap()

    /** 收到的好友请求列表  */
    private val friendRequests: Int2ObjectMap<FriendRequest> = Int2ObjectOpenHashMap()

    fun preload(request: QueryPlayerRequest) {
        request.preload(FriendshipData::class.java)
        request.preload(FriendRequest::class.java)
    }

    fun unpack(dataInfo: DataInfo) {
        for (friendship in dataInfo.unpackList(FriendshipData::class.java)) {
            friendships.put(friendship.friendId, friendship)
        }
        for (friendRequest in dataInfo.unpackList(FriendRequest::class.java)) {
            friendRequests.put(friendRequest.friendId, friendRequest)
        }
    }

    fun pack(dataInfo: DataInfo) {
        for (friendship in friendships.values) {
            dataInfo.pack(friendship)
        }
        for (friendRequest in friendRequests.values) {
            dataInfo.pack(friendRequest)
        }
    }

    fun init() {
//        player.registerOfflineHandler(moduleName, this::handleOfflineData)
    }

    val friends: Collection<FriendshipData>
        get() = Collections.unmodifiableCollection(friendships.values)

    fun isFriend(friendId: Int): Boolean {
        return friendships.containsKey(friendId)
    }

    fun getFriend(friendId: Int): FriendshipData {
        return friendships[friendId]
    }

    val friendCount: Int
        get() = friendships.count()

    fun getFriendRequests(): Collection<FriendRequest> {
        return Collections.unmodifiableCollection(friendRequests.values)
    }

    fun hasFriendRequest(friendId: Int): Boolean {
        return friendRequests.containsKey(friendId)
    }

    fun getFriendRequest(playerId: Int): FriendRequest {
        return friendRequests[playerId]
    }

    fun deleteFriend(friendId: Int, friendDbId: Int) {
        val friendship = friendships.remove(friendId) ?: return
        player.delete(friendship)
        pushOfflineData(friendId, friendDbId, FriendshipOfflineData.DELETE, player.playerId, player.dbId, 0L)
    }

    fun rejectFriend(friendId: Int) {
        val friendRequest = friendRequests.remove(friendId) ?: return
        player.delete(friendRequest)
    }

    fun acceptFriend(friendId: Int, friendDbId: Int) {
        val friendRequest = friendRequests.remove(friendId) ?: return
        player.delete(friendRequest)
        var friendship = friendships[friendId]
        if (friendship != null) return
        friendship = FriendshipData()
        friendship.friendId = friendId
        friendship.createTime = now()
        friendships.put(friendId, friendship)
        player.insert(friendship)
        pushOfflineData(
            friendId,
            friendDbId,
            FriendshipOfflineData.ACCEPT,
            player.playerId,
            player.dbId,
            friendship.createTime
        )
    }

    fun addFriend(friendId: Int, friendDbId: Int) {
        if (friendships.containsKey(friendId)) return
        pushOfflineData(
            friendId,
            friendDbId,
            FriendshipOfflineData.REQUEST,
            player.playerId,
            player.dbId,
            now()
        )
    }

    // 通知请求不要在这里处理，要单独发
    private fun handleOfflineData(json: String, isOnline: Boolean): Boolean {
        val data: FriendshipOfflineData = gson.fromJson(json, FriendshipOfflineData::class.java)
        when (data.cmd) {
            FriendshipOfflineData.REQUEST -> {
                if (friendRequests.containsKey(data.friendId)) return true
                val friendRequest = FriendRequest()
                friendRequest.friendId = data.friendId
                friendRequest.createTime = data.createTime
                friendRequests.put(friendRequest.friendId, friendRequest)
                player.insert(friendRequest)
            }

            FriendshipOfflineData.ACCEPT -> {
                var friendship = friendships[data.friendId]
                if (friendship != null) return true
                friendship = FriendshipData()
                friendship.friendId = data.friendId
                friendship.createTime = data.createTime
                friendships.put(friendship.friendId, friendship)
                player.insert(friendship)
            }

            FriendshipOfflineData.DELETE -> {
                val friendship = friendships.remove(data.friendId) ?: return true
                player.delete(friendship)
            }

            else -> {
                return false
            }
        }
        return true
    }

    private fun pushOfflineData(friendId: Int, friendDbId: Int, cmd: Int, playerId: Int, dbId: Int, createTime: Long) {
        val offlineData = FriendshipOfflineData(cmd, playerId, dbId, createTime)
        val json: String = gson.toJson(offlineData)
        player.pushOfflineDataTo(friendId, friendDbId, moduleName, json)
    }
}
